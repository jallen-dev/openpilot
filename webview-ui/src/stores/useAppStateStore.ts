import { create } from "zustand"

interface AppState {
  workspaceIndexed: boolean
  setWorkspaceIndexed: (workspaceIndexed: boolean) => void
}

export const useAppStateStore = create<AppState>()((set, get) => ({
  workspaceIndexed: false,
  setWorkspaceIndexed: (workspaceIndexed: boolean) => {
    set({ workspaceIndexed })
  }
}))
