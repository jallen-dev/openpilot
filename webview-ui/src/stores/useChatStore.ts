import { create } from "zustand"
import { immer } from "zustand/middleware/immer"

import { ChatMessage } from "../../../src/shared/ChatMessage"
import { vscode } from "../utils/vscode"

interface ChatState {
  userPrompt: string
  loading: boolean // used to show loading indicator for non-streaming chats
  chatMessages: ChatMessage[]
  setUserPrompt: (prompt: string) => void
  setLoading: (loading: boolean) => void
  setChatMessages: (messages: ChatMessage[]) => void
}

export const useChatStore = create<ChatState>()(
  immer((set, get) => ({
    userPrompt: vscode.getState()?.userPrompt ?? "",
    chatMessages: vscode.getState()?.chatMessages ?? [],
    loading: false,
    setUserPrompt: (userPrompt: string) => {
      vscode.setState({ userPrompt })
      set({ userPrompt })
    },
    setLoading: (loading: boolean) => set({ loading }),
    setChatMessages: (chatMessages: ChatMessage[]) => {
      vscode.setState({ chatMessages })
      set({ chatMessages })
    }
  }))
)
