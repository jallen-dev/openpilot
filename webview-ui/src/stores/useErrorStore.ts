import { create } from "zustand"

import { vscode } from "../utils/vscode"

interface ErrorState {
  errorMessage: string
  setErrorMessage: (errorMessage: string) => void
}

export const useErrorStore = create<ErrorState>()((set, get) => ({
  errorMessage: vscode.getState()?.errorMessage ?? "",
  setErrorMessage: (errorMessage: string) => {
    vscode.setState({ errorMessage })
    set({ errorMessage })
  }
}))
