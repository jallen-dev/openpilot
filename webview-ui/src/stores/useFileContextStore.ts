import { create } from "zustand"

import { File, FileSearchResult } from "../../../src/shared/File"
import { vscode } from "../utils/vscode"

interface FileContextState {
  suggestedFiles: FileSearchResult[]
  rejectedFilePaths: string[]
  setSuggestedFiles: (files: FileSearchResult[]) => void
  rejectFile: (file: string) => void
  undoRejectFile: (file: string) => void
  clearRejectedFiles: () => void
  getFilesToInclude: () => File[]
}

export const useFileContextStore = create<FileContextState>()((set, get) => ({
  suggestedFiles: vscode.getState()?.suggestedFiles ?? [],
  rejectedFilePaths: vscode.getState()?.rejectedFilePaths ?? [],
  setSuggestedFiles: (suggestedFiles: FileSearchResult[]) => {
    vscode.setState({ suggestedFiles })
    set({ suggestedFiles })
  },
  rejectFile: (file: string) =>
    set((state) => {
      vscode.setState({ rejectedFilePaths: [...state.rejectedFilePaths, file] })
      return {
        rejectedFilePaths: [...state.rejectedFilePaths, file]
      }
    }),
  undoRejectFile: (file: string) =>
    set((state) => {
      vscode.setState({
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      })
      return {
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      }
    }),
  clearRejectedFiles: () => {
    vscode.setState({ rejectedFilePaths: [] })
    set({ rejectedFilePaths: [] })
  },
  getFilesToInclude: () => {
    const { suggestedFiles, rejectedFilePaths } = get()
    return suggestedFiles.filter((f) => !rejectedFilePaths.includes(f.path))
  }
}))
