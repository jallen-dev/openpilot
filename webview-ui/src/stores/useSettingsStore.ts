import { create } from "zustand"

import { ApiKeys } from "../../../src/shared/ApiKeys"
import { Configuration } from "../../../src/shared/Configuration"
import { Model } from "../../../src/shared/Model"

interface SettingsState {
  model: Model
  apiKeys: ApiKeys
  includeFiles: boolean
  produceDiffs: boolean
  setModel: (model: Model) => void
  setApiKeys: (apiKeys: ApiKeys) => void
  setIncludeFiles: (includeFiles: boolean) => void
  setProduceDiffs: (produceDiffs: boolean) => void
  setConfiguration: (configuration: Configuration) => void
}

export const useSettingsStore = create<SettingsState>()((set, get) => ({
  model: "gpt-3.5-turbo",
  apiKeys: {},
  includeFiles: false,
  produceDiffs: false,
  setModel: (newModel) => set({ model: newModel }),
  setApiKeys: (newKeys) => {
    const { apiKeys } = get()
    set({ apiKeys: { ...apiKeys, ...newKeys } })
  },
  setIncludeFiles: (includeFiles) => set({ includeFiles }),
  setProduceDiffs: (produceDiffs) => set({ produceDiffs }),
  setConfiguration: (configuration: Configuration) => {
    const { includeFiles, produceDiffs } = configuration
    set({ includeFiles, produceDiffs })
  }
}))
