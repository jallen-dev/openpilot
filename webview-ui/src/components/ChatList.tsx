import { ReactMarkdown } from "react-markdown/lib/react-markdown"
import remarkGfm from "remark-gfm"

import { ApiKeys } from "../../../src/shared/ApiKeys"
import {
  ChatMessage,
  ModelMessage,
  UserMessage
} from "../../../src/shared/ChatMessage"
import {
  Model,
  OPENPILOT_MODEL,
  displayNameForModel
} from "../../../src/shared/Model"
import useMessageListener from "../hooks/useMessageListener"
import { useAppStateStore } from "../stores/useAppStateStore"
import { useChatStore } from "../stores/useChatStore"
import { useErrorStore } from "../stores/useErrorStore"
import { useSettingsStore } from "../stores/useSettingsStore"
import { SyntaxHighlighterWrapper } from "./SyntaxHighlighterWrapper"

function ChatList() {
  const chatMessages = useChatStore((state) => state.chatMessages)
  const setChatMessages = useChatStore((state) => state.setChatMessages)

  const workspaceIndexed = useAppStateStore((state) => state.workspaceIndexed)

  const model = useSettingsStore((state) => state.model)
  const apiKeys = useSettingsStore((state) => state.apiKeys)

  const errorMessage = useErrorStore((state) => state.errorMessage)

  const loading = useChatStore((state) => state.loading)

  useMessageListener("clearChat", () => {
    setChatMessages([])

    useErrorStore.getState().setErrorMessage("")
  })

  return (
    <section className="w-full h-full">
      {[
        startingChat({ model, workspaceIndexed, apiKeys }),
        ...chatMessages
      ].map((chat, index) => (
        <div key={index}>
          {isUserChat(chat) ? <UserChat {...chat} /> : <ModelChat {...chat} />}
        </div>
      ))}
      {loading && <div className="p-4">Waiting for response...</div>}
      {errorMessage && (
        <div className="p-4 bg-red-800 text-white">{errorMessage}</div>
      )}
    </section>
  )
}

export default ChatList

function UserChat({ content }: UserMessage) {
  return (
    <div className="p-4 bg-vscode-textBlockQuote-background border-t border-b border-zinc-500">
      <span>You: </span>
      <span>{content}</span>
    </div>
  )
}

function ModelChat({ content, participant }: ModelMessage) {
  return (
    <div className="p-4 overflow-x-scroll">
      <span>{displayNameForModel(participant)}: </span>
      <ReactMarkdown
        children={content}
        remarkPlugins={[remarkGfm]}
        components={{
          code({ node, inline, className, children, ...props }) {
            const match = /language-(\w+)/.exec(className || "")
            return !inline && match ? (
              <SyntaxHighlighterWrapper
                {...props}
                children={String(children).replace(/\n$/, "")}
                language={match[1]}
              />
            ) : (
              <code {...props} className={className}>
                {children}
              </code>
            )
          }
        }}
      />
    </div>
  )
}

function isUserChat(chat: UserMessage | ModelMessage): chat is UserMessage {
  return chat.participant === "user"
}

function startingChat({
  model,
  workspaceIndexed,
  apiKeys
}: {
  model: Model
  workspaceIndexed: boolean
  apiKeys: ApiKeys
}): ChatMessage {
  if (!(apiKeys.OpenAI || apiKeys.Google)) {
    return {
      content: `Welcome to OpenPilot! You first need to set your API keys.\n\nClick the ••• menu in the top right of this panel and select "Set API Key."\n\nTo create a new OpenAI key, visit https://platform.openai.com/account/api-keys`,
      participant: OPENPILOT_MODEL
    }
  }

  if (!apiKeys.OpenAI) {
    return {
      content: `In order for OpenPilot to automatically include relevant files for context, you first need to index your workspace. This requires an API key from OpenAI.\n\nYou can still chat with the LLM without this feature enabled.`,
      participant: model
    }
  }

  if (!workspaceIndexed) {
    return {
      content: `In order for OpenPilot to automatically include relevant files for context, you first need to index your workspace. Select "Index Workspace" from the menu in the top right.\n\nYou can still chat with the LLM without this feature enabled.`,
      participant: model
    }
  }

  return {
    content: `Hello! Type your prompt and OpenPilot will automatically include any relevant files into the chat context.
  
  Switch chat models using the menu in the top right.`,
    participant: model
  }
}
