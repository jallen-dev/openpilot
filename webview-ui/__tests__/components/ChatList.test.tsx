import { screen } from "@testing-library/dom"
import { render } from "@testing-library/react"

import ChatList from "../../src/components/ChatList"
import { useChatStore } from "../../src/stores/useChatStore"

test("renders just the starting message when there's no chats", () => {
  render(<ChatList />)

  expect(screen.getByText(/welcome/i)).toBeInTheDocument()
})

test("renders user chats and model chats", () => {
  useChatStore.getState().setChatMessages([
    { content: "user chat", participant: "user" },
    { content: "model chat", participant: "gpt-3.5-turbo" }
  ])

  render(<ChatList />)

  expect(screen.getByText(/user chat/i)).toBeInTheDocument()
  expect(screen.getByText(/model chat/i)).toBeInTheDocument()
})

test("renders markdown in model chats", () => {
  useChatStore.getState().setChatMessages([
    {
      content: "# Heading\n\n```js\nconst test = 'hello world'\n```",
      participant: "gpt-3.5-turbo"
    }
  ])

  render(<ChatList />)

  expect(screen.getByText("Heading").tagName).toBe("H1")
  expect(document.querySelector("code")).toBeInTheDocument()
})

test("matches snapshot", () => {
  useChatStore.getState().setChatMessages([
    { content: "user chat", participant: "user" },
    { content: "model chat", participant: "gpt-3.5-turbo" }
  ])

  const { asFragment } = render(<ChatList />)

  expect(asFragment()).toMatchSnapshot()
})
