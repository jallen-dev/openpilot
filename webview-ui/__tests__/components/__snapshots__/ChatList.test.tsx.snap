// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="w-full h-full"
  >
    <div>
      <div
        class="p-4 overflow-x-scroll"
      >
        <span>
          OpenPilot: 
        </span>
        <p>
          Welcome to OpenPilot! You first need to set your API keys.
        </p>
        

        <p>
          Click the ••• menu in the top right of this panel and select "Set API Key."
        </p>
        

        <p>
          To create a new OpenAI key, visit 
          <a
            href="https://platform.openai.com/account/api-keys"
          >
            https://platform.openai.com/account/api-keys
          </a>
        </p>
      </div>
    </div>
    <div>
      <div
        class="p-4 bg-vscode-textBlockQuote-background border-t border-b border-zinc-500"
      >
        <span>
          You: 
        </span>
        <span>
          user chat
        </span>
      </div>
    </div>
    <div>
      <div
        class="p-4 overflow-x-scroll"
      >
        <span>
          GPT 3.5 Turbo: 
        </span>
        <p>
          model chat
        </p>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
